### PR TITLE
Closes #2289: Add basic linkchecker extractor for az_card, and az_accordion field types.

### LINKED ISSUE #2289: Allow az_card and az_accordion field types to be checked with the link checker module.
## Motivation
Some sites use the link checker module to check if they have broken links.  This is generally a good practice, so we should support it.

### Is your feature request related to a problem? Please describe.
I can enable link checker for all fields in quickstart except az_card and az_accordion

## Proposed Resolution
Add a link extractor plugin for those field types.

### Describe the solution you'd like
I would like to be able to use link_checker with all fields.

### Describe alternatives you've considered
Front-end link checking  with [linkinator](https://github.com/JustinBeckwith/linkinator) or [dubbot](https://dubbot.com/features/broken-link-checks.html).

### PR DESCRIPTION
**Pull request by trackleft**
## Description
Adding a basic linkchecker extractor that looks in the link_url field of value array.

Could likely do the same for the `body` field, copying https://git.drupalcode.org/project/linkchecker/-/blob/8.x-1.x/src/Plugin/LinkExtractor/HtmlLinkExtractor.php

## Related issues
Closes #2289 

## How to test
Use this PR as a patch on an existing site with linkchecker on it.

``` 
composer config extra.patches.az-digital/az_quickstart --json '{"Quickstart issue #2289": "https://patch-diff.githubusercontent.com/raw/az-digital/az_quickstart/pull/2291.diff"}'
```

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by joeparsons — APPROVED** [review]

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ *     "az_card",
+ *     "az_accordion",
+ *   }
+ * )
+ */
+class AzLinkExtractor extends LinkExtractorBase {
+
+  /**
+   * {@inheritdoc}
+   */
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:24]: Does this method exist in the parent class?

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ * @LinkExtractor(
+ *   id = "az_link_extractor",
+ *   label = @Translation("Quickstart field link extractor"),
+ *   field_types = {
+ *     "az_card",
+ *     "az_accordion",
+ *   }
+ * )
+ */
+class AzLinkExtractor extends LinkExtractorBase {
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:20]:
```suggestion
class AzLinkExtractor extends HtmlLinkExtractor {
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    return $urls;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function extractUrlFromField(array $value) {
+
+    $link_uri = $this->extractUrlFromLinkUriField($value);
+    $body = $this->extractUrlFromBodyField($value);
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:162]:
```suggestion
    $body = parent::extractUrlFromField($value);
    $link_uri = $this->extractUrlFromLinkUriField($value);
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+   */
+  protected function extractUrlFromLinkUriField(array $value) {
+    // Return the uri index from the $value array.
+    return empty($value['link_uri']) ? [] : [$value['link_uri']];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function extractUrlFromBodyField(array $value) {
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:33]: Should be possible to remove this method with the suggested changes.

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ *     "az_accordion",
+ *   }
+ * )
+ */
+class AzLinkExtractor extends HtmlLinkExtractor {
+
+
+  /**
+   * Extracts a URLs from field.
+
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:24]:
```suggestion
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ *   id = "az_link_extractor",
+ *   label = @Translation("Quickstart field link extractor"),
+ *   field_types = {
+ *     "az_card",
+ *     "az_accordion",
+ *   }
+ * )
+ */
+class AzLinkExtractor extends HtmlLinkExtractor {
+
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:20]:
```suggestion
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  /**
+   * Extracts a URLs from field.
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:22]:
```suggestion
   * Extracts a URLs from field.
   *
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+   * @param array $value
+   *   The field value.
+   *
+   * @return array
+   *   Array of URLs.
+   */
+   protected function extractUrlFromLinkUriField(array $value) {
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:29]:
```suggestion
  protected function extractUrlFromLinkUriField(array $value) {
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    // Return the uri index from the $value array.
+    return empty($value['link_uri']) ? [] : [$value['link_uri']];
+  }
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:32]:
```suggestion
   }
```

File: modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+   * @param array $value
+   *   The field value.
+   *
+   * @return array
+   *   Array of URLs.
+   */
+  protected function extractUrlFromLinkUriField(array $value) {
+    // Return the uri index from the $value array.
+    return empty($value['link_uri']) ? [] : [$value['link_uri']];
+   }
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/LinkExtractor/AzLinkExtractor.php:33]:
```suggestion
  }
```

**Comment by trackleft** [comment]
Tested this in Campus Web Services' prelaunch module that we plan on using prior to launching sites.
Appears to work as expected.

**Review by tadean — APPROVED** [review]

**Pull request merged**